File: Spring-Boot/jpa-udemy/src/main/java/dev/vitorvidal/jpaudemy/repository/StudentRepository.java
package dev.vitorvidal.jpaudemy.repository;

import dev.vitorvidal.jpaudemy.entity.Passport;
import dev.vitorvidal.jpaudemy.entity.Student;
import jakarta.persistence.EntityManager;
import jakarta.transaction.Transactional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Repository;

@Repository
@Transactional
public class StudentRepository {
    private final Logger logger = LoggerFactory.getLogger(this.getClass());
    @Autowired
    private EntityManager entityManager;

    public Student findById(Long id) {
        return entityManager.find(Student.class, id);
    }

    public void deleteById(Long id) {
        Student student = findById(id);
        entityManager.remove(student);
    }

    public Student save(Student student) {
        if (student.getId() == null) {
            entityManager.persist(student);
            return student;
        } else {
            return entityManager.merge(student);
        }
    }

    public void saveStudentWithPassport() {
        Passport passport = new Passport("4985");
        entityManager.persist(passport);
        Student student = new Student("Arianne");
        student.setPassport(passport);
        entityManager.persist(student);
    }

    public void playWithEntityManager() {
        logger.info("playWithEntityManager - start");
        Student student = new Student("Web services student");
        entityManager.persist(student);
        student.setName("Web services student updated");
    }
}

File: Spring-Boot/jpa-kotlin-example/src/main/kotlin/dev/vitorvidal/jpakotlinexample/repository/StudentRepository.kt
package dev.vitorvidal.jpakotlinexample.repository

import dev.vitorvidal.jpakotlinexample.entity.Passport
import dev.vitorvidal.jpakotlinexample.entity.Student
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Repository
import javax.persistence.EntityManager
import javax.transaction.Transactional

@Repository
@Transactional
class StudentRepository {
    private val logger = LoggerFactory.getLogger(this.javaClass)

    @Autowired
    private val entityManager: EntityManager? = null
    fun findById(id: Long?): Student {
        return entityManager!!.find(Student::class.java, id)
    }

    fun deleteById(id: Long?) {
        val student = findById(id)
        entityManager!!.remove(student)
    }

    fun save(student: Student): Student {
        return if (student.id == null) {
            entityManager!!.persist(student)
            student
        } else {
            entityManager!!.merge(student)
        }
    }

    fun saveStudentWithPassport() {
        val passport = Passport(null, "4985", null, null, null)
        entityManager!!.persist(passport)
        val student = Student(null, "Arianne", passport, null, null)
        student.passport = passport
        entityManager.persist(student)
    }

    fun playWithEntityManager() {
        logger.info("playWithEntityManager - start")
        val student = Student(null, "Web services student", null, null, null)
        entityManager!!.persist(student)
        student.name = "Web services student updated"
    }
}
